src/backend/feed/processor.js

```diff
@@ -127,10 +127,15 @@ module.exports = async function processor(job) {
 
   let info;
   const invalid = await feed.isInvalid();
+  const delayed = await feed.isDelayed();
```

Comment:
You can let these two Promises run at the same time vs. blocking the second on the first:

```js
const [invalid, delayed] = await Promise.all(feed.isInvalid(), feed.isDelayed());
```

Reply:
Hmm, something's up that's causing `npm test` (specifically, `feed-processor.test.js`) to fail locally when I try to use `Promise.all()`:

<details>
<summary>(Click to expand <code>npm test</code> log)</summary>

raungar@ubuntu:~/repos/telescope$ npm test> @seneca/telescope@0.6.0 pretest /home/raungar/repos/telescope
> npm run lint

> @seneca/telescope@0.6.0 lint /home/raungar/repos/telescope
> npm run eslint


> @seneca/telescope@0.6.0 eslint /home/raungar/repos/telescope
> eslint --ignore-path .gitignore .


/home/raungar/repos/telescope/src/backend/feed/processor.js
  129:36  warning  Promise.all() requires 1 argument, but received 2  promise/valid-params

✖ 1 problem (0 errors, 1 warning)


> @seneca/telescope@0.6.0 postlint /home/raungar/repos/telescope
> npm run stylelint


> @seneca/telescope@0.6.0 stylelint /home/raungar/repos/telescope
> stylelint src/**/*.css


> @seneca/telescope@0.6.0 test /home/raungar/repos/telescope
> npm run jest


> @seneca/telescope@0.6.0 jest /home/raungar/repos/telescope
> cross-env LOG_LEVEL=error MOCK_REDIS=1 jest --

 FAIL  test/feed-processor.test.js
  ● Feed Processor Tests › Passing a valid Atom feed URI should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      15 |     fixtures.nockValidAtomResponse();
      16 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 17 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      18 |   });
      19 | 
      20 |   test('Passing a valid RSS feed URI should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:17:11)

  ● Feed Processor Tests › Passing a valid RSS feed URI should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      23 |     fixtures.nockValidRssResponse();
      24 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 25 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      26 |   });
      27 | 
      28 |   test('Passing a valid URI with HTML response should work', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:25:11)

  ● Feed Processor Tests › Passing a valid URI with HTML response should work

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      31 |     fixtures.nockValidHtmlResponse();
      32 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 33 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      34 |   });
      35 | 
      36 |   test('Passing an invalid RSS category feed should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:33:11)

  ● Feed Processor Tests › Passing an invalid RSS category feed should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      39 |     fixtures.nockInvalidRssResponse();
      40 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 41 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      42 |   });
      43 | 
      44 |   test('Passing a valid RSS category feed should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:41:11)

  ● Feed Processor Tests › Passing a valid RSS category feed should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      47 |     fixtures.nockValidRssResponse();
      48 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 49 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      50 |   });
      51 | 
      52 |   test('Non existent feed failure case: 404 should work', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:49:11)

  ● Feed Processor Tests › Non existent feed failure case: 404 should work

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      55 |     fixtures.nock404Response();
      56 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 57 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      58 |   });
      59 | });
      60 | 

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:57:11)

 PASS  test/posts.test.js
 PASS  test/post.test.js
 PASS  test/basic_analysis.test.js
 PASS  test/feeds.test.js
 PASS  test/feeds-atom.test.js
 PASS  test/feeds-opml.test.js
 PASS  test/feeds-wiki.test.js
 PASS  test/spam-checker.test.js
 PASS  test/feeds-rss.test.js
 PASS  test/server.test.js
 PASS  test/feeds-json.test.js
 PASS  test/extract-urls.test.js
 PASS  test/summarize.test.js
 PASS  test/wiki-feed-parser.test.js
[ 2020-02-16 13:08:40.371 ] ERROR (36232 on ubuntu): Unable to download wiki feed data from url https://wiki.cdot.senecaco
llege.ca/wiki/Planet_CDOT_Feed_List
    error: {}
[ 2020-02-16 13:08:40.373 ] ERROR (36232 on ubuntu): Unable to download wiki feed data from url https://wiki.cdot.senecaco
llege.ca/wiki/Planet_CDOT_Feed_List
    error: {}
 PASS  test/sanitize-html.test.js
 PASS  test/text-parser.test.js
 PASS  test/feed-queue.test.js
 PASS  test/storage.test.js
(node:36232) UnhandledPromiseRejectionWarning: Error: expect(received).resolves.toBe(expected) // Object.is equality

Expected: true
Received: false
(node:36232) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing insid
e of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id
: 18)
(node:36232) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections 
that are not handled will terminate the Node.js process with a non-zero exit code.
 PASS  test/feed.test.js
 PASS  test/github.test.js
 PASS  test/count-blog-domain.test.js
 PASS  test/hash.test.js
 PASS  test/logger.test.js
 PASS  test/sentiment-analysis.test.js

Summary of all failing tests
 FAIL  test/feed-processor.test.js
  ● Feed Processor Tests › Passing a valid Atom feed URI should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      15 |     fixtures.nockValidAtomResponse();
      16 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 17 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      18 |   });
      19 | 
      20 |   test('Passing a valid RSS feed URI should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:17:11)

  ● Feed Processor Tests › Passing a valid RSS feed URI should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      23 |     fixtures.nockValidRssResponse();
      24 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 25 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      26 |   });
      27 | 
      28 |   test('Passing a valid URI with HTML response should work', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:25:11)

  ● Feed Processor Tests › Passing a valid URI with HTML response should work

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      31 |     fixtures.nockValidHtmlResponse();
      32 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 33 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      34 |   });
      35 | 
      36 |   test('Passing an invalid RSS category feed should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:33:11)

  ● Feed Processor Tests › Passing an invalid RSS category feed should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      39 |     fixtures.nockInvalidRssResponse();
      40 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 41 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      42 |   });
      43 | 
      44 |   test('Passing a valid RSS category feed should pass', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:41:11)

  ● Feed Processor Tests › Passing a valid RSS category feed should pass

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      47 |     fixtures.nockValidRssResponse();
      48 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 49 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      50 |   });
      51 | 
      52 |   test('Non existent feed failure case: 404 should work', async () => {

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:49:11)

  ● Feed Processor Tests › Non existent feed failure case: 404 should work

    expect(received).resolves.not.toBeDefined()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: object is not iterable (cannot read property Symbol(Symbol.iterator))]

      55 |     fixtures.nock404Response();
      56 |     const job = fixtures.createMockJobObjectFromFeedId(id);
    > 57 |     await expect(processor(job)).resolves.not.toBeDefined();
         |           ^
      58 |   });
      59 | });
      60 | 

      at expect (node_modules/expect/build/index.js:138:15)
      at Object.<anonymous> (test/feed-processor.test.js:57:11)


Test Suites: 1 failed, 24 passed, 25 total
Tests:       6 failed, 121 passed, 127 total
Snapshots:   0 total
Time:        7.932s, estimated 8s
Ran all test suites.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @seneca/telescope@0.6.0 jest: `cross-env LOG_LEVEL=error MOCK_REDIS=1 jest --`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @seneca/telescope@0.6.0 jest script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/raungar/.npm/_logs/2020-02-16T21_08_41_566Z-debug.log
npm ERR! Test failed.  See above for more details.
</details>

I noticed that ESLint isn't happy with this usage of `Promise.all` for some reason, but I don't think this is the cause of the issue:

![image](https://user-images.githubusercontent.com/13500769/74612759-53f4d880-50d6-11ea-8536-7d1f74c9ad10.png)

(Also, in case it might have helped, I did also try adding additional `await` statements, e.g. `const [invalid, delayed] = await Promise.all(await feed.isInvalid(), await feed.isDelayed());`, but ended up with the same result and tests failing.)

Not sure how to proceed to figure this one out.

Reply:
You need to adjust your call to `Promise.all()`.  The way it works, is you give it an `Array` of `Promise` objects.  You won't `await` the result on each, just return the `Promise`, and let `Promise.all()` consolidate these into a single `Promise` that you `await` just once:

```js
const [invalid, delayed] = await Promise.all([feed.isInvalid(), feed.isDelayed()]);
```

Reply:
Ah, that worked! I was under the mistaken impression that `Promise.all()` would concatenate its arguments into an array.

Done: https://github.com/Seneca-CDOT/telescope/pull/733/commits/093fcef0926b7971ed32e20092b5dbb742d5a035